0017-sampler-interface.md

```diff
@@ -0,0 +1,282 @@
+In the case of the `Sampler`, a task consists only of a circuit and an array of parameter value sets if it is a parametric circuit.
+We aim to move the `Sampler` to be a viable replacement for `backend.run`, and eliminate the notion of sample weights (i.e. quasi-probabilities).
+
+## Motivation <a name="motivation"></a>
+
+We will propose a `Sampler` that
+
+ * _Obeys program outputs._ Every circuit defines a set of classical registers whose output names and types are known before execution. The prototypical type is a bit-array, but the OpenQASM 3 specification allows many other types, too, which Qiskit aims to follow. The current return structure of the sampler does not handle this gracefully, squishing bitstrings from separate registers together. It also does not attempt make a careful distinction between sampling from a program's output registers, and sampling from the output state that it prepares---it did not need to as these two views are equivalent for circuits that measure all qubits at the end of the circuit. However, this proposal takes the former stance, and separates returned data according to its output names. The motivation for this choice is that it would make the `Sampler` fully compliant with dynamic circuits.
+ * _Naturally allows multiple parameter value sets._ A common use case for many users is running the same parametric program with many parameter value sets in the same `run()` invocation. These changes will make it easier for implementations to do this efficiently with late-stage binding. Efficient invocation of the same parametric circuit between multiple `run()` invocations is out of scope of this RFC.
+ * _Is a "`memory=True`" exection model._ We will break from the tradition of using a mapping structure as the default container type for circuit exection. Instead, results will always be returned in a format where one output axis iterates over shots, and maintains execution order. The motivation is that, in a program whose total output size is more than 100 qubits, due to gate and measurement noise, the chances that any two output values are equal to each other is too small to make special accomodation for via a hash map. When the output size is smaller than this loose threshhold, so that output value collisions are likely, the data will take up less space anyways. Container objects will contain methods returning count-like structures to ease migration.
+ * _Has no sample weights._ The current sampler attaches real weights summing to one to all samples. In effect, it returns an estimate of a (quasi)-distribution instead of samples from the distribution. This proposal eliminates this feature and returns only samples, and without weights. The rationale is to make the `Sampler` into more of a streamlined circuit executor, paring down all frills.
```

Comment:
If an implementation that does those things is willing to pretend that it returns just samples from the output, then I guess so. However, I think the spirit of the proposal is to recognize those techniques as non-scalable and therefore out of scope of the Sampler. Someone willing to accept finite-size limitations from ROM can construct a quasi-prob distribution client-side.

Reply:
I think M3 readout error mitigation was ostensibly scalable. And I remember one reason we said users should change their programs to use Primitives rather than `backend.run()` was that they get access to error mitigation done server side. In the case of readout error mitigation, it is much easier for the runtime to hold and update the readout error model, than for each user to characterize this separately.

I think the data type returned from "mitigated counts" might not be compatible with pure samples as it's not a string of `0` and `1`. It might be possible for those methods to reconstruct a (quasi)-probability distribution, and then resample from that to generate "fake" samples and pretend those are the real samples. I don't know if this is the correct way of going about this. 

I don't want to hold up the RFC more, but I think this requires some more thought.

Reply:
@ajavadia there very well may be users that want an interface on quasi-probability distributions. However, the methods we have today don't scale to utility-scale examples. I would happy to discuss ways to resolve that in a separate issue.

Reply:
Imagine a device on 1000 qubits where each qubit has a 0.01 readout error. Consider a circuit that does nothing but measure each qubit once, into a classical register on 1000 bits. In 100,000 shots, the chances of observing the all-0 bitstring are about 4e-5, and the chances of observing the all-0 bitstring in 10K shots at least once are about 35%. Consider that these numbers go quickly to 0 as a function of both circuit width and depth, keeping instruction error fixed.

M3 doesn't change the observed bitstrings, it only re-weights their likelihood away from 1/N. So in effect, the above is showing that, asymptotically, M3 is guaranteed to report a quasi-prob distribution that is disjoint from the circuit's ideal distribution. One might be able to argue that this is okay, because one is ultimately not interested in length-1000 bitstrings, but small marginals thereof. This would require careful messaging.

Keep in mind also that quadratic scaling in circuit size is a problem in practice, because eventually post-processing

> it is much easier for the runtime to hold and update the readout error model, than for each user to characterize this separately

Are there other existing readout mitigation methods, besides M3, where this actually works? Because the other ones I know about explode to 2^n in memory. You can get around this by only doing mitigation on marginalized subsystems, but this kind of thing makes more sense client side, because otherwise you need to introduce semantics about how and where to marginalize, and lose the ability to change your mind having taken the data.

> returned from "mitigated counts" might not be compatible with pure samples as it's not a string of 0 and 1

We considered adding a field to the result bundle called `weights` which would give real weights to the corresponding samples. This would effectively make the output equivalent to a multimap, where the keys are some index of the sample dimension, and the values are the weights. This could be added in the future if a mitigation procedure for which it makes sense becomes available.